File: ios/Classes/FlutterIzipayPlugin.swift
import Flutter
import IzipayPayButtonSDK
import MastercardSonic
import UIKit
import VisaSensoryBranding

/// Flutter plugin that bridges the Izipay iOS SDK.
///
/// Registered automatically via CocoaPods. Do not instantiate directly.
public class FlutterIzipayPlugin: NSObject, FlutterPlugin, IzipayPaymentDelegate {
  private var eventSink: FlutterEventSink?
  private let sonicManager: MCSonicController = MCSonicController()

  public static func register(with registrar: FlutterPluginRegistrar) {
    let methodChannel = FlutterMethodChannel(
      name: "izipay_sdk/method_channel",
      binaryMessenger: registrar.messenger())
    let eventChannel = FlutterEventChannel(
      name: "izipay_sdk/event_channel",
      binaryMessenger: registrar.messenger())

    let instance = FlutterIzipayPlugin()
    registrar.addMethodCallDelegate(instance, channel: methodChannel)
    eventChannel.setStreamHandler(instance)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    // Force Light Mode — the Izipay form is not designed for Dark Mode.
    UIApplication.shared.windows.first?.rootViewController?
      .overrideUserInterfaceStyle = .light

    guard let args = call.arguments as? [String: String] else {
      result(FlutterError(code: "INVALID_ARGS",
                          message: "Expected [String: String] arguments",
                          details: nil))
      return
    }

    switch call.method {
    case "payWithCard":
      payWithCard(data: args)
    case "payWithYape":
      payWithYape(data: args)
    case "openFormToSaveCard":
      result(FlutterError(code: "UNIMPLEMENTED",
                          message: "openFormToSaveCard is not yet supported on iOS",
                          details: nil))
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  // MARK: - Payment methods

  private func payWithCard(data: [String: String]) {
    let config = buildConfig(data: data, payMethods: [.card], action: "pay")
    presentPaymentForm(configPayment: config)
  }

  private func payWithYape(data: [String: String]) {
    let config = buildConfig(data: data, payMethods: [.yape_code], action: "pay")
    presentPaymentForm(configPayment: config)
  }

  // MARK: - Config builder

  private func buildConfig(
    data: [String: String],
    payMethods: [PaymentMethodIzipay],
    action: String
  ) -> ConfigPaymentIzipay {
    let timestampMs = Int(Date().timeIntervalSince1970 * 1000)

    var config = ConfigPaymentIzipay()
    config.enviroment = data["environment"]
    config.action = action
    // SDK v2.3.0 renamed `publicKey` to `authorization` internally.
    config.authorization = data["publicKey"]
    config.transactionId = data["transactionId"]
    config.merchantCode = data["merchantCode"]

    config.order = OrderPaymentIzipay()
    config.order?.orderNumber = data["orderNumber"]
    config.order?.currency = "PEN"
    config.order?.amount = data["amount"]
    config.order?.payMethod = payMethods
    config.order?.processType = "authorize"
    config.order?.merchantBuyerId = data["userId"]
    config.order?.dateTimeTransaction = "\(timestampMs)"

    config.token = TokenPaymentIzipay()

    config.billing = BillingPaymentIzipay()
    config.billing?.firstName = data["firstName"]
    config.billing?.lastName = data["lastName"]
    config.billing?.email = data["email"]
    config.billing?.phoneNumber = data["phoneNumber"]
    config.billing?.street = data["street"]
    config.billing?.city = data["city"]
    config.billing?.state = data["state"]
    config.billing?.country = data["country"]
    config.billing?.postalCode = data["postalCode"]
    config.billing?.documentType = data["documentType"]
    config.billing?.document = data["documentNumber"]

    config.shipping = ShippingPaymentIzipay()

    config.appearance = AppearencePaymentIzipay()
    config.appearance?.language = "ESP"
    config.appearance?.formControls = AppearenceControlsPaymentIzipay()
    config.appearance?.formControls?.isAmountLabelVisible = true
    config.appearance?.formControls?.isLangControlVisible = false
    config.appearance?.visualSettings = AppearenceVisualSettingsPaymentIzipay()
    config.appearance?.visualSettings?.presentationMode = .fullscreen
    config.appearance?.theme = "purple"
    config.appearance?.customTheme = CustomThemePaymentIzipay()
    config.appearance?.logo = data["logoUrl"]

    let visaBranding = SensoryBranding()
    visaBranding.backdropColor = .white
    visaBranding.isSoundEnabled = true
    visaBranding.isHapticFeedbackEnabled = true
    visaBranding.checkmarkMode = .checkmark

    let mastercardBranding = MCSonicView()
    mastercardBranding.background = MCSonicBackground.white

    config.appearance?.sensoryBrand = AppearenceSensoryBrandIzipay()
    config.appearance?.sensoryBrand?.visaSBView = visaBranding
    config.appearance?.sensoryBrand?.mastercardSBView = mastercardBranding

    config.urlIPN = data["webhookUrl"]

    return config
  }

  // MARK: - Presentation

  private func presentPaymentForm(configPayment: ConfigPaymentIzipay) {
    guard
      let rootViewController = UIApplication.shared.connectedScenes
        .compactMap({ $0 as? UIWindowScene })
        .flatMap({ $0.windows })
        .first(where: { $0.isKeyWindow })?.rootViewController
    else {
      sendResultToFlutter(result: [
        "success": false,
        "error": "Could not resolve root view controller",
      ])
      return
    }

    let storyboard = UIStoryboard(
      name: "IziPayment",
      bundle: Bundle(for: PaymentFormViewController.self))
    let vc = storyboard.instantiateViewController(
      withIdentifier: "PaymentFormView") as! PaymentFormViewController
    vc.delegate = self
    vc.configPayment = configPayment
    vc.modalPresentationStyle = .fullScreen

    DispatchQueue.main.async {
      rootViewController.present(vc, animated: true)
    }
  }

  // MARK: - IzipayPaymentDelegate

  public func getPaymentResult(_ paymentResult: PaymentResult) {
    let code = paymentResult.code ?? ""
    if code == "00" {
      sendResultToFlutter(result: [
        "success": true,
        "code": code,
        "cardToken": paymentResult.response?.token?.cardToken as Any,
        "cardPan": paymentResult.response?.card?.pan as Any,
        "cardBrand": paymentResult.response?.card?.brand as Any,
      ])
    } else {
      sendResultToFlutter(result: [
        "success": false,
        "code": code,
        "message": paymentResult.message as Any,
      ])
    }
  }

  public func executeProfiling(_ result: IzipayPayButtonSDK.ScoringParams) {
    // ThreatMetrix profiling callback — no action required from the plugin.
  }

  public func executeSensoryBrandVisa(view: UIView) {
    guard let visaView = view as? SensoryBranding else { return }
    DispatchQueue.main.asyncAfter(deadline: .now() + 1.0) {
      visaView.animate()
    }
  }

  public func executeSensoryBrandMastercard(
    view: UIView, params: IzipayPayButtonSDK.MastercardSonicParams
  ) {
    guard
      let mastercardView = view as? MCSonicView,
      let merchant = MCSonicMerchant(
        merchantName: params.name,
        countryCode: params.country,
        city: nil,
        merchantCategoryCodes: [params.mcc],
        merchantId: nil)
    else { return }

    sonicManager.prepare(
      with: .soundAndAnimation,
      sonicCue: params.sonicCue,
      sonicEnvironment: .production,
      merchant: merchant,
      isHapticsEnabled: true
    ) { [weak self] status in
      guard status == .success else { return }
      self?.sonicManager.play(with: mastercardView) { _ in }
    }
  }
}

// MARK: - FlutterStreamHandler

extension FlutterIzipayPlugin: FlutterStreamHandler {
  public func onListen(
    withArguments arguments: Any?,
    eventSink events: @escaping FlutterEventSink
  ) -> FlutterError? {
    self.eventSink = events
    return nil
  }

  public func onCancel(withArguments arguments: Any?) -> FlutterError? {
    self.eventSink = nil
    return nil
  }

  public func sendResultToFlutter(result: [String: Any?]) {
    eventSink?(result)
  }
}

File: android/src/main/kotlin/dev/darkcodex/izipay_sdk/FlutterIzipayPlugin.kt
package dev.darkcodex.izipay_sdk

import android.app.Activity
import android.content.Intent
import android.util.Log

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken

import com.izipay.izipay_pw_sdk.data.model.ConfigRequest
import com.izipay.izipay_pw_sdk.data.model.OrderPaymentIzipay
import com.izipay.izipay_pw_sdk.data.model.PayOption
import com.izipay.izipay_pw_sdk.data.model.TokenPaymentIzipay
import com.izipay.izipay_pw_sdk.data.model.BillingPaymentIzipay
import com.izipay.izipay_pw_sdk.data.model.AppearencePaymentIzipay
import com.izipay.izipay_pw_sdk.data.model.AppearenceControlsPaymentIzipay
import com.izipay.izipay_pw_sdk.data.model.AppearenceVisualSettingsPaymentIzipay
import com.izipay.izipay_pw_sdk.data.model.CustomThemePaymentIzipay
import com.izipay.izipay_pw_sdk.domain.model.PaymentResponse
import com.izipay.izipay_pw_sdk.ui.fullscreend.ContainerActivity

class FlutterIzipayPlugin : FlutterPlugin, MethodCallHandler, EventChannel.StreamHandler, ActivityAware {
    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var eventSink: EventChannel.EventSink? = null
    private var activity: Activity? = null

    companion object {
        /** Request code used to identify the Izipay activity result. */
        private const val REQUEST_CODE_IZIPAY = 1001
        private const val IZIPAY_ACTION_PAY = "pay"
        private const val IZIPAY_ACTION_REGISTER = "register"
        /** Intent extra key expected by ContainerActivity. */
        private const val EXTRA_REQUEST = "REQUEST"
        /** Intent extra key containing the JSON response from the SDK. */
        private const val EXTRA_RESPONSE_PAYLOAD = "RESPONSEJSON"
        private const val TAG = "IzipayPlugin"
        /** Izipay success response code. */
        private const val SUCCESS_CODE = "00"
    }

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel = MethodChannel(binding.binaryMessenger, "izipay_sdk/method_channel")
        methodChannel.setMethodCallHandler(this)
        eventChannel = EventChannel(binding.binaryMessenger, "izipay_sdk/event_channel")
        eventChannel.setStreamHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
        binding.addActivityResultListener { requestCode, resultCode, data ->
            handleActivityResult(requestCode, resultCode, data)
            true
        }
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        @Suppress("UNCHECKED_CAST")
        val data = call.arguments as? Map<String, String> ?: run {
            result.error("INVALID_ARGS", "Expected Map<String, String>", null)
            return
        }
        when (call.method) {
            "payWithCard" -> launchPayment(data, PayOption.CARD, IZIPAY_ACTION_PAY)
            "payWithYape" -> launchPayment(data, PayOption.YAPE, IZIPAY_ACTION_PAY)
            "openFormToSaveCard" -> launchPayment(data, PayOption.CARD, IZIPAY_ACTION_REGISTER)
            else -> result.notImplemented()
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    private fun launchPayment(map: Map<String, String>, payOption: PayOption, action: String) {
        val currentActivity = activity ?: run {
            Log.e(TAG, "Activity is null — cannot launch payment")
            sendEvent(mapOf("success" to false, "error" to "Activity not available"))
            return
        }

        val isRegister = action == IZIPAY_ACTION_REGISTER
        val amount = if (isRegister) "0.00" else map["amount"] ?: ""

        val request = ConfigRequest(
            map["environment"] ?: "",
            map["publicKey"] ?: "",
            action,
            map["transactionId"] ?: "",
            map["merchantCode"] ?: "",
            "",
            OrderPaymentIzipay(
                map["orderNumber"] ?: "",
                "PEN",
                amount,
                arrayListOf(payOption),
                "authorize",
                map["userId"] ?: "",
                System.currentTimeMillis().toString(),
            ),
            TokenPaymentIzipay(""),
            BillingPaymentIzipay(
                map["firstName"] ?: "",
                map["lastName"] ?: "",
                map["email"] ?: "",
                map["phoneNumber"] ?: "",
                map["street"] ?: "",
                map["city"] ?: "",
                map["state"] ?: "",
                map["country"] ?: "",
                map["postalCode"] ?: "",
                map["documentType"] ?: "",
                map["documentNumber"] ?: "",
            ),
            null,
            AppearencePaymentIzipay(
                "ESP",
                // Show the pay button for payments; hide it for card registration.
                AppearenceControlsPaymentIzipay(!isRegister, false),
                AppearenceVisualSettingsPaymentIzipay(true),
                "purple",
                CustomThemePaymentIzipay("#333399", "#333399", "#333399"),
                map["logoUrl"] ?: "",
            ),
            map["webhookUrl"] ?: "",
            null,
            null,
        )

        val intent = Intent(currentActivity, ContainerActivity::class.java)
        intent.putExtra(EXTRA_REQUEST, request)
        currentActivity.startActivityForResult(intent, REQUEST_CODE_IZIPAY)
    }

    private fun handleActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode != REQUEST_CODE_IZIPAY) return

        val rawJson = data?.getStringExtra(EXTRA_RESPONSE_PAYLOAD)

        if (rawJson.isNullOrEmpty()) {
            Log.w(TAG, "No response payload from SDK (resultCode=$resultCode)")
            sendEvent(mapOf("success" to false, "error" to "No response from SDK"))
            return
        }

        try {
            val response: PaymentResponse = Gson().fromJson(rawJson, object : TypeToken<PaymentResponse>() {}.type)
            sendEvent(mapOf(
                "success" to (response.code == SUCCESS_CODE),
                "code" to response.code,
                "message" to response.message,
                "cardToken" to response.response?.token?.cardToken,
                "cardPan" to response.response?.card?.pan,
                "cardBrand" to response.response?.card?.brand,
            ))
        } catch (e: Exception) {
            Log.e(TAG, "Failed to parse PaymentResponse: ${e.message}")
            sendEvent(mapOf("success" to false, "error" to "Parse error: ${e.message}"))
        }
    }

    private fun sendEvent(data: Map<String, Any?>) {
        eventSink?.success(data)
    }
}
